api/client/application/client_test.go

```diff
@@ -439,58 +486,79 @@ func (s *applicationSuite) TestDestroyUnits(c *gc.C) {
 			DetachedStorage:  []params.Entity{{Tag: "storage-pgdata-1"}},
 		},
 	}}
+	result := new(params.DestroyUnitResults)
+	results := params.DestroyUnitResults{expectedResults}
 	delay := 1 * time.Minute
-	client := newClient(func(objType string, version int, id, request string, a, response interface{}) error {
-		c.Assert(request, gc.Equals, "DestroyUnit")
-		c.Assert(a, jc.DeepEquals, params.DestroyUnitsParams{
-			Units: []params.DestroyUnitParams{
-				{UnitTag: "unit-foo-0", Force: true, MaxWait: &delay},
-				{UnitTag: "unit-bar-1", Force: true, MaxWait: &delay},
-			},
-		})
-		c.Assert(response, gc.FitsTypeOf, &params.DestroyUnitResults{})
-		out := response.(*params.DestroyUnitResults)
-		*out = params.DestroyUnitResults{expectedResults}
-		return nil
-	})
-	results, err := client.DestroyUnits(application.DestroyUnitsParams{
+
+	args := params.DestroyUnitsParams{
+		Units: []params.DestroyUnitParams{
+			{UnitTag: "unit-foo-0", Force: true, MaxWait: &delay},
+			{UnitTag: "unit-bar-1", Force: true, MaxWait: &delay},
+		},
+	}
+	mockFacadeCaller := mocks.NewMockFacadeCaller(ctrl)
+	mockFacadeCaller.EXPECT().FacadeCall("DestroyUnit", args, result).SetArg(2, results).Return(nil)
+
+	client := application.NewClientFromCaller(mockFacadeCaller)
+	res, err := client.DestroyUnits(application.DestroyUnitsParams{
 		Units:   []string{"foo/0", "bar/1"},
 		Force:   true,
 		MaxWait: &delay,
 	})
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(results, jc.DeepEquals, expectedResults)
+	c.Assert(res, jc.DeepEquals, expectedResults)
 }
 
 func (s *applicationSuite) TestDestroyUnitsArity(c *gc.C) {
-	client := newClient(func(objType string, version int, id, request string, a, response interface{}) error {
-		return nil
-	})
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	result := new(params.DestroyUnitResults)
+	args := params.DestroyUnitsParams{
+		Units: []params.DestroyUnitParams{
+			{UnitTag: names.NewUnitTag("foo/0").String()},
+		},
+	}
+	mockFacadeCaller := mocks.NewMockFacadeCaller(ctrl)
+	mockFacadeCaller.EXPECT().FacadeCall("DestroyUnit", args, result).Return(errors.New("expected 1 result(s), got 0"))
```

Comment:
The "expected 1 results(s), not 0" does result from the mock result, it should come from api code when the number of request does not equal the number of results received, see line 471.